### ISSUE TITLE
Builder Image Scripts: Ease to Maintain, Test, and Trust

### ISSUE THREAD

**Issue by rhcarvalho** [openshift/source-to-image#421]
For the lack of a central repository holding all of our official builder images, and for `sti-base` not representing our images equally well (it is not used by database images), I'm bringing the discussion here, for here is the source of a builder image skeleton: `s2i create`.
## Proposal

Look for another programming language for writing scripts in the official builder images to improve maintainability, including the ability to write more and better tests and gain more trust on the scripts' behavior.

tldr; let's call that language **Python**
## Rationale

All scripting today in our official images is done in Bash, and it has proven itself not suitable for the task of writing not-so-trivial scripts.

Some common problems we see:
- Not all scripts set safe shell options (`errexit`, `pipefail`, `nounset`, ...)
- Bash "gotchas" like declaring and defining a `local` variable in the same line
- Unhandled error conditions
- Poor ability to decompose tasks as reusable pieces / functions
- Very poor "testing framework"
- I don't have much more hair left to lose

According to a [Shell Style Guide by Google](https://google.github.io/styleguide/shell.xml?showone=When_to_use_Shell#When_to_use_Shell), that I've been often pointed to while working on our images:

> Shell should only be used for small utilities or simple wrapper scripts.
> ...
> - If you find you need to use arrays for anything more than assignment of ${PIPESTATUS}, you should use Python.
> - If you are writing a script that is more than 100 lines long, you should probably be writing it in Python instead. Bear in mind that scripts grow. Rewrite your script in another language early to avoid a time-consuming rewrite at a later date.

I feel we went way way way way past the line where Bash can do the trick.

We've collectively wasted too much time trying to tame Bash to do what it is not particularly good at doing.
## Another Language?

I keep three constraints in mind when thinking about proposing a language change:
- Is it available in the base Operating System or would we need to install extra packages?
- Do we have know-how in the team to use another language equally or better than Bash?
- How do we go about "migrating everything"? There are cost involved there.

From the top of my head, there are 2 environments that comes with a CentOS/RHEL base install:
- Bash
- **Python**

Perl was almost in the list, but I realized it is not in a base install (and we don't have it in the MongoDB image, for example).

I believe the SCL team has the required Python know-how.

How we go about a potential migration, if we decided to do so, can be discussed here. I wouldn't try to keep everything in sync as we have tried. A big-bang migration today is impractical, too expensive. FWIW we are already in a situation where we have discrepancies in the "script framework" of the images.

We could start with rewriting tests for one of the images, see how it goes.
## Images Involved

Databases:
- https://github.com/openshift/mongodb
- https://github.com/openshift/mysql
- https://github.com/openshift/postgresql

Base image for language builders:
- https://github.com/openshift/sti-base

Language builders:
- https://github.com/openshift/sti-nodejs
- https://github.com/openshift/sti-perl
- https://github.com/openshift/sti-php
- https://github.com/openshift/sti-python
- https://github.com/openshift/sti-ruby

**Comment by rhcarvalho** [comment]
@bparees @mfojtik @mnagy @php-coder @PI-Victor @legionus @hhorak @ppisar @omron93 what do you think?

Are you all happy with the current state and the Bash scripts?

(feel free to tag other SCL image maintainers for the discussion)

**Comment by rhcarvalho** [comment]
I would open a separate issue for this, but I'm thinking this might be influenced by the language choice:

**We have a growing technical debt with copy-paste of Bash scripts supporting multiple versions of a given image.**

For example, the Python image has 3 copies of mostly the same files with a few differences between each copy.

Perhaps one of the causes of that state is that it would be "too hard" to make the scripts work for multiple versions using Bash scripts.

Perhaps with Python it would have been much easier and a no-brainer to just make a single script have say 3 different implementations for a certain function and pick the right one according to the image version, while keeping no duplication across the code base.

And all that with better (unit) tests to prevent regressions.

**Comment by mfojtik** [comment]
@rhcarvalho the tests you're talking about are executing `s2i build`, `docker run` and inspecting the result by looking into container logs or sending HTTP requests against container. I don't think we have tests that are more complex than that.

So I appreciate your effort to make this better, but I really don't want to spend time on rewriting all image tests to python (or any other language framework). I would be OK if SCL folks (@hhorak) are wiling to spend time on it (and maintain them).

Other problem I see here is that we will require upstream/community to write a "python" test for new features, which might be an unnecessary barrier for them.

> Perhaps one of the causes of that state is that it would be "too hard" to make the scripts work for multiple versions using Bash scripts.

I don't see how it will be better in Python comparing to Bash. You can make functions in Bash that will do the same job as the Python functions. Also, can you name the differences? Is it something we cover in Bash functions as arguments? 

(and for the record, I'm not saying Bash is the greatest testing framework, it sux big time, but I don't think having the tests in Python will sux less ;-)

**Comment by ppisar** [comment]
Arguing by having something in default installation is flawed. First you don't have the docker in the default installation, yet it's not a hurdle from installing it and using it. Second, future default installation will miss even the Python with all it's standard classes probably.

But whatever you choose,  it will be better than bash. Shell is terribly bad in calling foreign code. I'd like to see a testing library provided by s2i (or something else out of the image source to prevent from copy-and-pasting followed by divergence and rotting) that could ease writing tests for the images. And I mean good test library that will provide test aggregation, detailed failure reporting, parallel testing etc.

**Comment by PI-Victor** [comment]
If we're thinking long term it would probably have to be python 3 to begin with, or maybe i'm thinking too long therm, but anyway:
```
The End Of Life date (EOL, sunset date) for Python 2.7 has been moved five years into the future, to 2020. This decision was made to clarify the status of Python 2.7 and relieve worries for those users who cannot yet migrate to Python 3. See also PEP 466.
```
+1 on the idea, i don't see it happening, you're going to hit the ol' "it's not broken so we don't fix it" wall. i hope you prove me wrong!
bash pretty much sucks for anything this complex. i have no idea why we would want to maintain this level of bash scripts in such an environment that will grow and change even more when online is going to be operational.

**Comment by rhcarvalho** [comment]
> Arguing by having something in default installation is flawed. First you don't have the docker in the default installation, yet it's not a hurdle from installing it and using it.

This is true for running tests in our development machines, but not true for the runtime scripts that run in Docker under CentOS and RHEL images.

If it was only about tests, we would be free to use whatever language out there.

Since the scripts must run in a Docker image and we don't want to make the images unnecessarily bigger, thus the need for something that is already there (or a very good reason to install something else to all official images).

**Comment by mfojtik** [comment]
@rhcarvalho you don't have to include the scripts in the image (and I don't think that will make them bigger). You can just copy them to a running container and execute them there.

**Comment by rhcarvalho** [comment]
> @rhcarvalho you don't have to include the scripts in the image (and I don't think that will make them bigger). You can just copy them to a running container and execute them there.

Hmmm, I don't get it. Maybe we are thinking about different things?

I'm talking about these, for example:

- https://github.com/openshift/mongodb/blob/master/2.4/root/usr/bin/run-mongod (script part of the image's runtime)
- https://github.com/openshift/mongodb/blob/master/2.4/test/run (test script)

Obviously the runtime scripts need to be in the image. The test scripts are a different matter.

The common thing between the two is that they are written in Bash and share the problems I listed.


What would make the image bigger is not a few kilobytes of scripts, but using a language/framework that is not part of the base OS (CentOS/RHEL) install. For example, one could propose writing everything in Rust. That would require installing extra packages in the base image, increasing the image size for all images we're supporting.

Does that make sense?

**Comment by ppisar** [comment]
On Thu, Mar 03, 2016 at 02:51:38AM -0800, Rodolfo Carvalho wrote:
> I'm talking about these, for example:
> 
> - https://github.com/openshift/mongodb/blob/master/2.4/root/usr/bin/run-mongod (script part of the image's runtime)
> - https://github.com/openshift/mongodb/blob/master/2.4/test/run (test script)
> 
I thought we talk only about the test scripts. The test scripts are not part
of the resulting image. Therefore I cannot understand why a non-developer
would need need to run them in a pristine RHEL/CentOS container. Or is that
another use case of the tests? This why I asked for the constrains for the
tests.

**Comment by mfojtik** [comment]
@rhcarvalho so you're talking about running the test scripts inside the container (which in case they are written in python will require python + modules needed to run them). 

There are couple use cases here for testing:

* Run database image and then docker exec into that container and use database client CLI to perform testing.
* Run database image and inspect the logs of the image to see if everything runs properly
* Run database image and inspect the content of the container (configuration file) to see if the environment variables used for configuring that container were used properly.
* Run database image and exercise the replication/clustering scenario by running multiple containers and mock the IP addresses to allow them to talk/see each-other
* Run s2i build using the language image and see if the build pass
* Run s2i build using the language image and then run the resulting application image and see if works
* Run s2i build using the language image and then run the resulting application image and inspect the container to see if everything was installed/configured properly

@rhcarvalho I think first, we should expand this list and see what exactly we have to test now. Then choose the appropriate framework (or stick with bash if we conclude that the bash is the best option).

Does it make sense?

**Comment by php-coder** [comment]
SGTM, but it takes time for rewriting and testing that's why I'm not sure that we'll do this.

**Comment by rhcarvalho** [comment]
> I thought we talk only about the test scripts.

I talk about all of the scripts, which I believe fall in two categories: test and runtime.

Clearly they go under different constraints, they run in different stages at different environments.

However, for me both suffer from having outgrow what we can maintain sanely with Bash.



> @rhcarvalho so you're talking about running the test scripts inside the container

No, I am not talking about running tests in containers.


> There are couple use cases here for testing... choose the appropriate framework (or stick with bash if we conclude that the bash is the best option).

In terms of "Turing completeness", both Bash and Python would be able to cover the points @mfojtik listed above.

Now Python would benefit from having a testing framework in the standard library, sane defaults (will stop on errors, functions have explicitly declared arguments, etc).

I'm not proposing writing Python + external dependencies nor anything fancy. Just stupid Python with the standard library, no external dependencies, just like now we don't bundle or import any Bash script from the Internet.



> SGTM, but it takes time for rewriting and testing that's why I'm not sure that we'll do this.

Yes, we're all aware of that cost.

Let's not forget the costs we're paying with debugging our Bash code base today. It is not there for free either. I don't think we're in the point of saying "it just works so let's not touch it", because we're constantly trying to fix problems and we spend time trying to fix them (see recent efforts with MongoDB, a lot of changes, and no tests).

**Comment by ppisar** [comment]
I do not expect rewriting existing tests forcefully. But I hope those who
extends current tests will find the new framework more appealing than the
hackish bash script and decide themselves to rewrite them.

**Comment by rhcarvalho** [comment]
@ppisar :+1: 

I've brought this up after having tried to improve scripts in the past and having certain changes rejected because:

1. "We need to maintain consistency across repositories, if you change the 'framework' you need to change it in all repos" https://github.com/openshift/sti-python/pull/12#issuecomment-100181427
2. "Bash is not a programming language, don't try to be smart with it" https://github.com/openshift/sti-python/pull/12#discussion_r31950199

It was almost a year ago, and instead of having a better state now, we have a much worse state where the consistency from (1) is long gone, and the complexity of our Bash code extrapolates my ability to reason about it (and successfully automate tests for my/our code).

If we collectively acknowledge that (1) is not a must have and that (2) yeah, Bash is not the language we should be using for all the scripts without considering alternatives, then I see a free path for anybody working on improvements to propose a change that can be accepted and merged and perhaps replicated later.

One such big rewrite work can be seen here: https://github.com/openshift/mongodb/pull/112

Unfortunately it's a Bash->Bash rewrite which today contains many of the Bash flaws we're been trying to fix here and there over the year. Anyway I support and respect @omron93 for his energy and courage to make bold improvements.

I'm not forcing or suggesting anybody we'll prioritize their time to go rewrite existing code.
But I want at least a written-down agreement that we need to change the current situation, if we can all get to that agreement.

The improvements would come over time, and this issue can serve as a reference point for pull requests that are not only tiny little changes to existing Bash scripts.

Am I over stating the problem? Do we all agree that the situation is far from ideal?

**Comment by bparees** [comment]
I'm pretty against this personally.  Perhaps because i have zero python experience, but bash is also an evolved and stable platform.  Are we going to be continually revising python scripts to keep up with new python releases and best practices?

Bash may be esoteric at times, but for a user looking to copy+modify something, it's a lower barrier to entry than python.  It's far more universal than python.

Look at origin for example...  all the tooling/test/build scripts are in bash.  I don't think a suggestion to origin (or k8s) to rewrite them in python would get very far.

Are there some special cases where the logic is getting complex enough that a more sophisticated scripting language might be appropriate (eg the mongo replication scripts?), yes, and i'd be more amenable to strategic changes in those places over a full switch everywhere.

And of course there's the practical argument:  we don't have the resources to rewrite this stuff and deal w/ the inevitable regressions.

so, tl;dr:
1) agree it's not ideal, but i think you're overstating it
2) bash is more universal and approachable (everyone knows bash, even if they don't know every bit of syntax.  not everyone knows python or wants to learn it)
3) amenable to strategic use of non-bash for particularly complicated bits.

**Comment by mfojtik** [comment]
agree with @bparees ... bash might be frustrating (i'm not arguing that), but it does the job for now and I would personally keep improving the scripts or derivate common parts into some "framework" we can reuse than committing us to rewrite to python/perl/whatever and risking we end up with the same maintenance hell ;-)

**Comment by clever88** [comment]
-1 on python too, as this proposal seems to be more of a personal preference than technical improvement.

**Comment by hhorak** [comment]
I'm also quite reluctant to invest in rewriting the testing scripts and I'm even more negative to rewriting images' internals. However, in case somebody tries it at least as PoC, I'd suggest to use already existing tools:
* for container testing: https://github.com/Containers-Testing-Framework/
* for container configuration: https://github.com/containers-tools/cct

**Comment by PI-Victor** [comment]
is this still up for debate, if not can we close it out?
my recent experience of using python on a project instead of bash wasn't any better so i'd like to change my previous statement in support of bash.

**Comment by bparees** [comment]
closing this since the SCL team owns those repos now and while they're welcome to rewrite the scripts in any language they choose, it doesn't seem likely.
